docs/design/thee_datahall.md

```diff
@@ -0,0 +1,197 @@
+Additionaly to these localties at least two different `locality_zone` per `data_hall`/`datacenter` is required to fullfill the requirements.
+The `locality_zone` can be configured with [FoundationDBClusterFaultDomain](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterfaultdomain).
+In addition to that the user can configure additional variables that the sidecar will use for substition in the [FoundationDBClusterSpec](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterspec) with the `sidecarVariables` list.
+The current implementation would allow to set `dataHall: $AZ` and in `sidecarVariables` we would list `AZ` to define the `locality_data_hall` based on an environment variable that will be replaced by the sidecar with the actual value.
+Instead of having these different mechanisims spread across different settings I propose to have a new `localities` setting in the `FoundationDBClusterSpec`:
+
+```yaml
+localities:
+- key: "locality_data_hall"
+  value: ""
+  valueFromEnv: ""
+  valueFromNode: "topology.kubernetes.io/zone"
+  topologyKey: ""
+- key: "locality_zone"
+  value: ""
+  valueFromEnv: "MyFancyZone"
+  valueFromNode: ""
+  topologyKey: "kubernetes.io/hostname"
+```
+
+The `topologyKey` will be used for `topologySpreadConstraints` and for `PodAntiAffinity` and only has to be set if `valueFromNode` is empty, otherwise it will default to that value.
+The `FoundationDBClusterFaultDomain` would then be reduced to only have `key`, `zoneCount` and `zoneIndex` since the locality will be configured in `localities`.
+For all `localities` that define a `valueFromEnv` we would add the key to the `--substitute-variable` flag.
+For `valueFromNode` we have to modify the sidecar to allow it to read labels from Kubernetes nodes (see [Related Links](#related-links)) and pass that information to the according new flag.
+This change should provide the most flexibility to the user to define the required/wanted localities.
+We would set `locality_instance_id` and `locality_machineid` to the current defaults but also allow the user to define custom localities.
+After that change we should deprecate `cluster.Spec.DataCenter` and `cluster.Spec.DataHall`.
+
+### Deployment model
+
+#### Single regional Kubernetes cluster
+
+The assumption here is that the user has a regional Kubernetes cluster that spreads across at least 3 different AZs.
+To ensure that all Pods are spread evenly across the AZs the operator should use [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints).
+
+```yaml
+  topologySpreadConstraints:
+  - maxSkew: 1
+    # the topologyKey will be read from the locality
+    topologyKey: topology.kubernetes.io/zone 
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        foundationdb.org/fdb-cluster-name: test-cluster
```

Comment:
I assume that the process classes needs to be evenly spread across zones:

```
      matchLabels:
        foundationdb.org/fdb-cluster-name: test-cluster
        fdb-process-class: <process-class>
```

Reply:
Yes that makes sense, seems like I forgot to add the process class label.